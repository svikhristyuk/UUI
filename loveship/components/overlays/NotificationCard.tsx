import React from 'react';
import { INotification, Icon, IHasChildren, IHasCX, UuiContext, UuiContexts } from '@epam/uui';
import * as styles from '../../assets/styles/scss/loveship-color-vars.scss';
import { IconContainer } from '@epam/uui-components';
import { IconButton } from '../buttons';
import { ReactComponent as SuccessIcon } from '../icons/notification-check-fill-24.svg';
import { ReactComponent as WarningIcon } from '../icons/notification-warning-fill-24.svg';
import { ReactComponent as ErrorIcon } from '../icons/notification-error-fill-24.svg';
import { ReactComponent as HintIcon } from '../icons/notification-help-fill-24.svg';
import { ReactComponent as CrossIcon } from '../icons/snackbar/cross.svg';
import * as css from './NotificationCard.scss';
import cx from 'classnames';
import { EpamColor, LinkButton } from '..';
import { i18n } from '../../i18n';

type notificationAction = {
    name: string;
    action: () => void;
};

export interface DefaultNotificationProps extends INotification, IHasChildren, IHasCX {
    actions?: notificationAction[];
}

export interface NotificationCardProps extends DefaultNotificationProps {
    icon?: Icon;
    color: EpamColor;
}

export class NotificationCard extends React.Component<NotificationCardProps> {
    notificationCardNode: HTMLElement = null;

    constructor(props: NotificationCardProps) {
        super(props);
        this.refNode = this.refNode.bind(this);
    }

    componentDidMount() {
        this.notificationCardNode.addEventListener('mouseenter', this.props.clearTimer);
        this.notificationCardNode.addEventListener('mouseleave', this.props.refreshTimer);
    }

    componentWillUnmount() {
        this.notificationCardNode.removeEventListener('mouseenter', this.props.clearTimer);
        this.notificationCardNode.removeEventListener('mouseleave', this.props.refreshTimer);
    }

    refNode (el: HTMLDivElement) {
        this.notificationCardNode = el;
    }

    render() {
        return <div className={ cx(css.notificationWrapper, styles[`color-${ this.props.color }`], this.props.cx) } ref={ this.refNode }>
            <div className={ css.mainPath }>
                {
                    this.props.icon && <div className={ css.iconWrapper }>
                        <IconContainer icon={ this.props.icon } cx={ css.actionIcon } />
                    </div>
                }
                <div className={ css.content }>
                    { this.props.children }
                    { this.props.actions && <div className={ css.actionWrapper }>
                        { this.props.actions.map((action: notificationAction) => {
                            return <LinkButton caption={ action.name } onClick={ action.action } key={ action.name } cx={ css.actionLink } size='36' />;
                        }) }
                    </div> }
                </div>
                { this.props.onClose && <IconButton icon={ CrossIcon } color='night600' onClick={ this.props.onClose } cx={ css.closeIcon } /> }
            </div>
        </div>;
    }
}

export const WarningNotification = (props: DefaultNotificationProps) => <NotificationCard icon={ WarningIcon } color='sun' { ...props } cx={ cx(props.cx, css.typeWarning) } />;
export const SuccessNotification = (props: DefaultNotificationProps) => <NotificationCard icon={ SuccessIcon } color='grass' { ...props } cx={ cx(props.cx, css.typeSuccess) } />;
export const HintNotification = (props: DefaultNotificationProps) => <NotificationCard icon={ HintIcon } color='sky' { ...props } cx={ cx(props.cx, css.typeHint) } />;
export const ErrorNotification = (props: DefaultNotificationProps) => <NotificationCard icon={ ErrorIcon } color='fire' { ...props } cx={ cx(props.cx, css.typeError) } />;

export class ClearNotification extends React.Component<{}> {
    public static contextType = UuiContext;
    public context: UuiContexts;

    render() {
        return <div className={ cx(css.notificationWrapper, css.clearButton) }><LinkButton caption={ i18n.notificationCard.closeAllNotificationsButton } onClick={ () => this.context.uuiNotifications.clearAll() } /></div>;
    }
}
