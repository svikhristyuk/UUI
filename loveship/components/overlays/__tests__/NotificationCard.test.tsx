import React from "react";
import renderer from "react-test-renderer";
import {ClearNotification, ErrorNotification, HintNotification, NotificationCard, SuccessNotification, WarningNotification} from "../NotificationCard";
import { ReactComponent as AcceptIcon } from "../../icons/accept-12.svg";

jest.spyOn(NotificationCard.prototype, "refNode")
    .mockImplementation(function (this: typeof NotificationCard.prototype) {
        this.notificationCardNode = {
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
        } as any;
    });

describe("NotificationCard", () => {
    it("should be rendered correctly", () => {
        const tree = renderer
            .create(<NotificationCard
                icon={ AcceptIcon }
                id={ 1 }
                key='test'
                color='sun'
                onClose={ jest.fn() }
                onSuccess={ jest.fn() }
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should be rendered correctly with icon", () => {
        const tree = renderer
            .create(<NotificationCard
                icon={ AcceptIcon }
                id={ 1 }
                key='test'
                color='sun'
                onClose={ jest.fn() }
                onSuccess={ jest.fn() }
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe("WarningNotification", () => {
    it("should be rendered correctly", () => {
        const tree = renderer
            .create(<WarningNotification
                id={ 1 }
                key='test'
                onClose={ jest.fn() }
                onSuccess={ jest.fn() }
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe("SuccessNotification", () => {
    it("should be rendered correctly", () => {
        const tree = renderer
            .create(<SuccessNotification
                id={ 1 }
                key='test'
                onClose={ jest.fn() }
                onSuccess={ jest.fn() }
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe("HintNotification", () => {
    it("should be rendered correctly", () => {
        const tree = renderer
            .create(<HintNotification
                id={ 1 }
                key='test'
                onClose={ jest.fn() }
                onSuccess={ jest.fn() }
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe("ErrorNotification", () => {
    it("should be rendered correctly", () => {
        const tree = renderer
            .create(<ErrorNotification
                id={ 1 }
                key='test'
                onClose={ jest.fn() }
                onSuccess={ jest.fn() }
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe("ClearNotification", () => {
    it("should be rendered correctly", () => {
        const tree = renderer
            .create(<ClearNotification key='test'/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});