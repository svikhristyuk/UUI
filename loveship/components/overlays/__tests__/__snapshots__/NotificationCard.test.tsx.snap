// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClearNotification should be rendered correctly 1`] = `
<div
  className="notificationWrapper clearButton"
>
  <button
    className="container uui-button-box uui-enabled -clickable root size-36 color-sky font-sans-semibold uui-no-left-icon uui-no-right-icon"
    href={null}
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="uui-caption line-height-18 font-size-14 v-padding-9"
    >
      CLOSE ALL NOTIFICATIONS
    </div>
  </button>
</div>
`;

exports[`ErrorNotification should be rendered correctly 1`] = `
<div
  className="notificationWrapper color-fire typeError"
>
  <div
    className="mainPath"
  >
    <div
      className="iconWrapper"
    >
      <div
        className="container uui-icon uui-enabled actionIcon"
        style={Object {}}
      >
        <SvgrURL
          className=""
        />
      </div>
    </div>
    <div
      className="content"
    />
    <button
      className="container uui-button-box uui-enabled -clickable root color-night600 closeIcon"
      href={null}
      onClick={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="container uui-icon uui-enabled"
        style={Object {}}
      >
        <SvgrURL
          className=""
        />
      </div>
    </button>
  </div>
</div>
`;

exports[`HintNotification should be rendered correctly 1`] = `
<div
  className="notificationWrapper color-sky typeHint"
>
  <div
    className="mainPath"
  >
    <div
      className="iconWrapper"
    >
      <div
        className="container uui-icon uui-enabled actionIcon"
        style={Object {}}
      >
        <SvgrURL
          className=""
        />
      </div>
    </div>
    <div
      className="content"
    />
    <button
      className="container uui-button-box uui-enabled -clickable root color-night600 closeIcon"
      href={null}
      onClick={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="container uui-icon uui-enabled"
        style={Object {}}
      >
        <SvgrURL
          className=""
        />
      </div>
    </button>
  </div>
</div>
`;

exports[`NotificationCard should be rendered correctly 1`] = `
<div
  className="notificationWrapper color-sun"
>
  <div
    className="mainPath"
  >
    <div
      className="iconWrapper"
    >
      <div
        className="container uui-icon uui-enabled actionIcon"
        style={Object {}}
      >
        <SvgrURL
          className=""
        />
      </div>
    </div>
    <div
      className="content"
    />
    <button
      className="container uui-button-box uui-enabled -clickable root color-night600 closeIcon"
      href={null}
      onClick={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="container uui-icon uui-enabled"
        style={Object {}}
      >
        <SvgrURL
          className=""
        />
      </div>
    </button>
  </div>
</div>
`;

exports[`NotificationCard should be rendered correctly with icon 1`] = `
<div
  className="notificationWrapper color-sun"
>
  <div
    className="mainPath"
  >
    <div
      className="iconWrapper"
    >
      <div
        className="container uui-icon uui-enabled actionIcon"
        style={Object {}}
      >
        <SvgrURL
          className=""
        />
      </div>
    </div>
    <div
      className="content"
    />
    <button
      className="container uui-button-box uui-enabled -clickable root color-night600 closeIcon"
      href={null}
      onClick={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="container uui-icon uui-enabled"
        style={Object {}}
      >
        <SvgrURL
          className=""
        />
      </div>
    </button>
  </div>
</div>
`;

exports[`SuccessNotification should be rendered correctly 1`] = `
<div
  className="notificationWrapper color-grass typeSuccess"
>
  <div
    className="mainPath"
  >
    <div
      className="iconWrapper"
    >
      <div
        className="container uui-icon uui-enabled actionIcon"
        style={Object {}}
      >
        <SvgrURL
          className=""
        />
      </div>
    </div>
    <div
      className="content"
    />
    <button
      className="container uui-button-box uui-enabled -clickable root color-night600 closeIcon"
      href={null}
      onClick={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="container uui-icon uui-enabled"
        style={Object {}}
      >
        <SvgrURL
          className=""
        />
      </div>
    </button>
  </div>
</div>
`;

exports[`WarningNotification should be rendered correctly 1`] = `
<div
  className="notificationWrapper color-sun typeWarning"
>
  <div
    className="mainPath"
  >
    <div
      className="iconWrapper"
    >
      <div
        className="container uui-icon uui-enabled actionIcon"
        style={Object {}}
      >
        <SvgrURL
          className=""
        />
      </div>
    </div>
    <div
      className="content"
    />
    <button
      className="container uui-button-box uui-enabled -clickable root color-night600 closeIcon"
      href={null}
      onClick={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="container uui-icon uui-enabled"
        style={Object {}}
      >
        <SvgrURL
          className=""
        />
      </div>
    </button>
  </div>
</div>
`;
